sklearn/tree/_tree.pyx

```diff
@@ -1230,3 +1228,399 @@ cdef class Tree:
+    """Record metrics used to return the cost complexity path"""
+    cdef DOUBLE_t[:] ccp_alphas
+    cdef DOUBLE_t[:] impurities
+    cdef UINT32_t count
+
+    def __cinit__(self,  int node_count):
+        self.ccp_alphas = np.zeros(shape=(node_count), dtype=np.float64)
+        self.impurities = np.zeros(shape=(node_count), dtype=np.float64)
+        self.count = 0
+
+    cdef int save_metrics(self,
+                          DOUBLE_t effective_alpha,
+                          DOUBLE_t subtree_impurities) nogil:
+        self.ccp_alphas[self.count] = effective_alpha
+        self.impurities[self.count] = subtree_impurities
+        self.count += 1
+
+
+cdef _cost_complexity_prune(
+    Tree orig_tree,
+    _CCPPruneController controller,
+    unsigned char[:] leaves_in_subtree,
+    SIZE_t[:] capacity):
+    """Performs cost complexity pruning.
+
+    The ccp_controller signals when the pruning should stop.
+
+    Parameters
+    ----------
+    orig_tree : Tree
+        Original tree
+    ccp_controller : _CCPPruneController
+        cost complexity controller
+    leaves_in_subtree : unsigned char[:]
+        output for leaves of subtree
+    capacity : SIZE_T*
+        output for capacity of subtree
+    """
+
+    cdef:
+        SIZE_t i
+        SIZE_t n_nodes = orig_tree.node_count
+        # prior probability using weighted samples
+        DOUBLE_t[:] weighted_n_node_samples = orig_tree.weighted_n_node_samples
+        DOUBLE_t total_sum_weights = weighted_n_node_samples[0]
+        DOUBLE_t[:] impurity = orig_tree.impurity
+        # weighted impurity of each node
+        DOUBLE_t[:] r_node = np.empty(shape=n_nodes, dtype=np.float64)
+
+        SIZE_t[:] child_l = orig_tree.children_left
+        SIZE_t[:] child_r = orig_tree.children_right
+        SIZE_t[:] parent = np.zeros(shape=n_nodes, dtype=np.intp)
+
+        # Only uses the start and parent variables
+        Stack stack = Stack(INITIAL_STACK_SIZE)
+        StackRecord stack_record
+        int rc = 0
+        SIZE_t node_idx
+
+        SIZE_t[:] n_leaves = np.zeros(shape=n_nodes, dtype=np.intp)
+        DOUBLE_t[:] r_branch = np.zeros(shape=n_nodes, dtype=np.float64)
+        DOUBLE_t current_r
+        SIZE_t leaf_idx
+        SIZE_t parent_idx
+
+        # candidate nodes that can be pruned
+        unsigned char[:] candidate_nodes = np.zeros(
+            shape=n_nodes, dtype=np.uint8)
+        # nodes in subtree
+        unsigned char[:] in_subtree = np.ones(shape=n_nodes, dtype=np.uint8)
+        DOUBLE_t[:] g_node = np.zeros(shape=n_nodes, dtype=np.float64)
+        SIZE_t pruned_branch_node_idx
+        DOUBLE_t subtree_alpha
+        DOUBLE_t effective_alpha
+        SIZE_t child_l_idx
+        SIZE_t child_r_idx
+        SIZE_t n_pruned_leaves
+        DOUBLE_t r_diff
+        DOUBLE_t max_float64 = np.finfo(np.float64).max
+
+    # find parent node ids and leaves
+    with nogil:
+
+        for i in range(r_node.shape[0]):
```

Comment:
do want to have `prange` in a place like here? Not sure how much it matters though, since the computation is pretty fast anyway.

Reply:
I have not played with opnemp in this PR. I do not think it is worth adding here.

Reply:
issue: https://github.com/scikit-learn/scikit-learn/issues/14037